internal/k8s/statuscache.go

```diff
@@ -0,0 +1,134 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	"fmt"
+
+	contour_api_v1 "github.com/projectcontour/contour/apis/projectcontour/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// StatusUpdateCacher takes status updates and applies them to a cache, to be used for testing.
+type StatusUpdateCacher struct {
+	objectCache map[string]interface{}
+}
+
+// IsCacheable returns whether this type of object can be stored in
+// the status cache.
+func (suc *StatusUpdateCacher) IsCacheable(obj interface{}) bool {
+	switch obj.(type) {
+	case *contour_api_v1.HTTPProxy:
+		return true
+	default:
+		return false
+	}
+}
+
+// OnDelete removes an object from the status cache.
+func (suc *StatusUpdateCacher) OnDelete(obj interface{}) {
+	if suc.objectCache != nil {
+		switch o := obj.(type) {
+		case *contour_api_v1.HTTPProxy:
+			delete(suc.objectCache, suc.objKey(o.Name, o.Namespace, contour_api_v1.HTTPProxyGVR))
+		default:
+			panic(fmt.Sprintf("status caching not supported for object type %T", obj))
+		}
+
+	}
+}
+
+// OnAdd adds an object to the status cache.
+func (suc *StatusUpdateCacher) OnAdd(obj interface{}) {
+	if suc.objectCache == nil {
+		suc.objectCache = make(map[string]interface{})
+	}
+
+	switch o := obj.(type) {
+	case *contour_api_v1.HTTPProxy:
+		suc.objectCache[suc.objKey(o.Name, o.Namespace, contour_api_v1.HTTPProxyGVR)] = obj
+		// fmt.Printf("Saved %s to cache, cache has %d entries\n", suc.objKey(o.Name, o.Namespace, contour_api_v1.HTTPProxyGVR), len(suc.objectCache))
+	default:
+		panic(fmt.Sprintf("status caching not supported for object type %T", obj))
```

Comment:
I still don't like the panic when an invalid type is encountered, not sure how others feel about but, but to me, doesn't seem like case where Contour should be terminated.

Reply:
If it needs to stay, should update the message to understand a bit more about where this was encountered.

Reply:
This is an assertion that programmers obeyed the invariants, so panic is OK by me.

Reply:
Yeah, the panic will only happen if a developer has passed the wrong type of thing, so this  should never happen. This is belt-and-suspenders territory.